darts/anomaly_detection/anomaly_model.py

```diff
@@ -0,0 +1,357 @@
+will train the model and/or the scorer, over the history of one time series. `score()` will apply the model on
+the time series input, and the scorer on the prediction of the model and the time series input. The `score()`
+will output the anomaly score of the time series input. The function `score_metric()` is the same as `score()`,
+but outputs the score of an agnostic threshold metric (AUC-ROC or AUC-PR), between the predicted anomaly score
+time series and a binary ground truth time series indicating the presence of anomalies.
+"""
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Sequence, Union
+
+from darts.anomaly_detection.score import Scorer
+from darts.logging import raise_if_not
+from darts.models.filtering.filtering_model import FilteringModel
+from darts.models.forecasting.forecasting_model import ForecastingModel
+from darts.timeseries import TimeSeries
+
+
+class AnomalyModel(ABC):
+    "Base class for all Anomaly Model"
+
+    @abstractmethod
+    def score(
+        self, series: Union[TimeSeries, Sequence[TimeSeries]]
+    ) -> Union[TimeSeries, Sequence[TimeSeries]]:
+        pass
+
+    @abstractmethod
+    def fit(
+        self, series: Union[TimeSeries, Sequence[TimeSeries]]
+    ) -> Union[TimeSeries, Sequence[TimeSeries]]:
+        pass
+
+    @abstractmethod
+    def score_metric(
+        self, series: Union[TimeSeries, Sequence[TimeSeries]]
+    ) -> Union[TimeSeries, Sequence[TimeSeries]]:
+        pass
+
+
+class ForecastingAnomalyModel(AnomalyModel):
+    def __init__(
+        self, model: ForecastingModel, scorer: Union[Scorer, Sequence[Scorer]]
+    ):
+        """Forecasting Anomaly Model
+
+        Parameters
+        ----------
+        model : ForecastingModel
+            A forecasting model from Darts that will be used to predict the actual time series
+        scorer : Scorer
+            A scorer that will be used to convert the actual and predicted time series to
+            an anomaly score time series. If a list of n scorer is given, the anomaly model will test each
+            one of the scorers and output n anomaly score.
+        """
+        super().__init__()
+
+        raise_if_not(
+            isinstance(model, ForecastingModel),
+            f"Model must be a darts.models.forecasting not a {type(model)}",
+        )
+        self.model = model
+
+        if isinstance(scorer, Sequence):
+            self.scorers = scorer
+        else:
+            self.scorers = [scorer]
+
+        for scorer in self.scorers:
+            raise_if_not(
+                isinstance(scorer, Scorer),
+                "Scorer must be a darts.anomaly_detection.score not a {}".format(
+                    type(scorer)
+                ),
+            )
+
+    def fit(
+        self,
+        series: TimeSeries,
+        model_fit_params: Optional[Dict[str, Any]] = None,
+        hist_forecasts_params: Optional[Dict[str, Any]] = None,
+    ):
```

Comment:
Subject for future bainstorming:

- For now we want to keep everything general, but isn't this hist_forecasts_params somehow a property of the scorer? And so, when we instantiate a scorer, we would have to input  a subset  of the `historical_forecast` parameters. Typically, the `forecast_horizon`, which. would give a prediction window on which we compute the metrics. 
- Also, where are the covariates ? (or maybe we have chosen to just do univariate for now, but in this case you have to check the type of model and that doesn't accept past and future covariates)
- Can we also have by default the fitting and the scoring training on the fitted model saved series, and then have series in input as optional?

Reply:
+1 with these 3 points. We should not expose `hist_forecasts_params`, but rather only expose whichever parameters are relevant to this particular `fit()` function (if any). Also we should accept covariates and multiple series here.